flocker/node/agents/ebs.py

```diff
@@ -0,0 +1,198 @@
+        """
+        Helper function to wait for up to 60s for given volume
+        to be in 'available' state.
+        """
+        start_time = time.time()
+        expected_volume.update()
+        while expected_volume.status != expected_status:
+            elapsed_time = time.time() - start_time
+            if elapsed_time < time_limit:
+                time.sleep(0.1)
+                expected_volume.update()
+            else:
+                raise Exception(
+                    'Timed out while waiting for volume. '
+                    'Expected Volume: {!r}, '
+                    'Expected Status: {!r}, '
+                    'Actual Status: {!r}, '
+                    'Elapsed Time: {!r}, '
+                    'Time Limit: {!r}.'.format(
+                        expected_volume, expected_status,
+                        expected_volume.status, elapsed_time,
+                        time_limit
+                    )
+                )
+
+    def _is_cluster_volume(self, cluster_id, ebs_volume):
+        """
+        Helper function to check if given volume belongs to
+        given cluster.
+        """
+        actual_cluster_id = ebs_volume.tags.get(CLUSTER_ID_LABEL)
+        if actual_cluster_id is not None:
+            actual_cluster_id = UUID(actual_cluster_id)
+            if actual_cluster_id == cluster_id:
+                return True
+        return False
+
+    def create_volume(self, dataset_id, size):
+        """
+        Create a volume on EBS. Store Flocker-specific
+        {metadata version, cluster id, dataset id} for the volume
+        as volume tag data.
+
+        :param String dataset_id: Dataset_id for the volume.
+        :param int size: Requested volume size in Bytes.
+
+        :return BlockDeviceVolume volume: Created volume in
+            BlockDeviceVolume format.
+        """
+        requested_volume = self.connection.create_volume(
+            size=int(Byte(size).to_GB().value), zone=self.zone)
+
+        # Stamp created volume with Flocker-specific tags.
+        metadata = {
+            METADATA_VERSION_LABEL: '1',
+            CLUSTER_ID_LABEL: unicode(self.cluster_id),
+            DATASET_ID_LABEL: unicode(dataset_id),
+        }
+        self.connection.create_tags([requested_volume.id],
+                                    metadata)
+
+        # Wait for created volume to reach 'available' state.
+        self._wait_for_volume(requested_volume)
+
+        # Return created volume in BlockDeviceVolume format.
+        return self._blockdevicevolume_from_ebs_volume(requested_volume)
+
+    def list_volumes(self):
+        """
+        Return all volumes in {available, in-use} state that belong to
+        this Flocker cluster.
+        """
+        volumes = []
+        for ebs_volume in self.connection.get_all_volumes():
+            if ((self._is_cluster_volume(self.cluster_id, ebs_volume)) and
+               (ebs_volume.status in [u'available', u'in-use'])):
```

Comment:
Why do we care about the status here?
What other status might they be in?
I'd have thought we want to return *all* volumes that appear to belong to the cluster and if their status is unexpected that indicates a bug in our code.
Oh, perhaps there's a "deleting" status...I haven't had to deal with that yet....let me read on.